### PR TITLE
8252015: [macos11] java.awt.TrayIcon requires updates for template images

### PR DESCRIPTION
**Pull request by petermz**
### Issue
[JDK-8252015: [macos11] java.awt.TrayIcon requires updates for template images](https://bugs.openjdk.java.net/browse/JDK-8252015)

### Problem
According to Apple's human interface guidelines, developers should use template images for tray icons. This way icons look good when desktop theme or wallpaper change. On the API level, this means setting the `NSImage::isTemplate` flag. Currently there's no way in Java to set this flag. Even if one uses a template (black and transparent pixels only) image, MacOS will treat it as regular image. As a result, the image will not be visible when using dark theme.

### Solution:
* Add a new system property, `java.awt.enableTemplateImages`, to indicate that template images are being used by tray icons.When this property is set, all tray icon images are treated as templates. It's the developers' responsibility to ensure their images are indeed templates. This property allows us to avoid new API and accidental behavior changes
* Value of this property is passed from Java into `NSImage::setTemplate`
* We need the tray icon image rendered by MacOS, so we use a button (rather than a view) to host the image

To do after CSR approval:
- [ ] Rename property `java.awt.`
- [ ] Add `TrayIcon` javadoc, use `@systemProperty` tag there

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252015](https://bugs.openjdk.java.net/browse/JDK-8252015): [macos11] java.awt.TrayIcon requires updates for template images

### Reviewers
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**)

### Contributors
 * Tres Finocchiaro `<tres.finocchiaro@gmail.com>`
 * Peter Zhelezniakov `<peterz@openjdk.org>`

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/481/head:pull/481`
`$ git checkout pull/481`

### CODE REVIEW

**Comment by petermz** [comment]
/contributor add Tres Finocchiaro <tres.finocchiaro@gmail.com>

**Comment by mrserb** [comment]
Hi, Peter. You are welcome!
Just an initial question about the template images, can you please provide some steps to reproduce this bug and so it will be possible to check that the fix solves it.

**Comment by petermz** [comment]
Hi Sergey, and thanks for looking into this!

Try the attached jar, it's a modified version of TrayIconDemo using a template icon taken from Wikimedia Commons. Without the fix, or without `-Dsun.awt.enableTemplateImages`, the icon disappears if you switch to Dark theme. With the fix and the property set, the system repaints it in white.

[TrayIconDemo.jar.zip](https://github.com/openjdk/jdk/files/5332241/TrayIconDemo.jar.zip)

**Comment by tresf** [comment]
> simply add a new comment to the pull request

✔️

**Comment by petermz** [comment]
Filed [CSR request](https://bugs.openjdk.java.net/browse/JDK-8255597)

File: src/java.desktop/macosx/native/libawt_lwawt/awt/CTrayIcon.m [hunk 1]
```diff
@@ -83,17 +99,23 @@ -(void) dealloc {
     // the item's view to nil: it can lead to a crash in some scenarios.
```
Comment (mrserb) [inline src/java.desktop/macosx/native/libawt_lwawt/awt/CTrayIcon.m:99]: This comment seems outdated, we either need to nullify the menu delegate or follow the old approach and update the comment on why the menuDelegate is not set to nil.

File: src/java.desktop/macosx/classes/sun/lwawt/macosx/CTrayIcon.java [hunk 2]
```diff
@@ -194,6 +194,8 @@ public void updateImage() {
     }
 
     void updateNativeImage(Image image) {
+        boolean imageTemplate = Boolean.parseBoolean(System.getProperty("sun.awt.enableTemplateImages", "false"));
```
Comment (mrserb) [inline src/java.desktop/macosx/classes/sun/lwawt/macosx/CTrayIcon.java:197]: "sun.awt.enableTemplateImages" -> "apple.awt.enableTemplateImages"

Comment (mrserb) [reply src/java.desktop/macosx/classes/sun/lwawt/macosx/CTrayIcon.java:197]: The impl block in the java.awt.TrayIcon class?

**Comment by mrserb** [comment]
Please check the tests in the .../test/jdk/java/awt/TrayIcon/ folder it looks like some of them fail afte the fix:
java/awt/TrayIcon/ActionCommand/ActionCommand.java 
java/awt/TrayIcon/ActionEventMask/ActionEventMask.java 
java/awt/TrayIcon/MouseEventMask/MouseEventMaskTest.java 
java/awt/TrayIcon/SecurityCheck/FunctionalityCheck/FunctionalityCheck.java 
java/awt/TrayIcon/TrayIconEventModifiers/TrayIconEventModifiersTest.java 
java/awt/TrayIcon/TrayIconEvents/TrayIconEventsTest.java 
java/awt/TrayIcon/TrayIconMouseTest/TrayIconMouseTest.java 
java/awt/TrayIcon/TrayIconPopup/TrayIconPopupTest.java

**Comment by petermz** [comment]
The icon delegate was not listening for mouse events -- thanks for catching this. I've restored the `AWTTrayIconView` thing (so that it can handle mouse events) and made it inherit from `NSStatusBarButton` (so that icon painting is delegated to MacOS). The fix is less verbose now. I've also renamed the property `apple.awt.enableTemplateImages` and added an `@implnote` as discussed in the CSR.

Can we consider the CSR approved btw?

**Comment by kevinrushforth** [comment]
@petermz no, the CSR is not in the Approved state. Once it is, the Skara bot will recognize this and remove the `csr` label such that it will not longer be an Integration Blocker.

**Comment by mrserb** [comment]
The new version of the changes looks much better, the test run is in progress.

**Review by mrserb — APPROVED** [review]

**Comment by voitylov** [comment]
@mrserb thank you for the review of the issue. Could you review the associated JDK-8255597 CSR as well?

**Comment by tresf** [comment]
@petermz do you mind editing the original bug report, incase this bug report is referenced in the future:

```diff
- Add a new system property, `java.awt.enableTemplateImages`, ...
+ Add a new system property, `apple.awt.enableTemplateImages`, ...
```

**Comment by petermz** [comment]
Sure, but I can't find this line in the bug report. Do you mean this one (https://bugs.openjdk.java.net/browse/JDK-8252015) ?

**Comment by tresf** [comment]
Sorry, I meant this PR's description, sorry for the confusion.

**Comment by petermz** [comment]
Done!

**Pull request closed without merging**